src/raw/mod.rs

```diff
@@ -0,0 +1,1634 @@
+//! A RawDocument can be created from a `Vec<u8>` containing raw BSON data, and elements
+//! accessed via methods similar to those available on the Document type. Note that rawbson returns
+//! a RawResult<Option<T>>, since the bytes contained in the document are not fully validated until
+//! trying to access the contained data.
+//!
+//! ```rust
+//! use bson::raw::{
+//!     RawBson,
+//!     RawDocument,
+//! };
+//!
+//! // See http://bsonspec.org/spec.html for details on the binary encoding of BSON.
+//! let doc = RawDocument::new(b"\x13\x00\x00\x00\x02hi\x00\x06\x00\x00\x00y'all\x00\x00".to_vec())?;
+//! let elem: Option<RawBson> = doc.get("hi")?;
+//!
+//! assert_eq!(
+//!   elem?.as_str()?,
+//!   "y'all",
+//! );
+//! # Ok::<(), bson::raw::RawError>(())
+//! ```
+//!
+//! ### bson-rust interop
+//!
+//! A [`RawDocument`] can be created from a [`bson::document::Document`]. Internally, this
+//! serializes the `Document` to a `Vec<u8>`, and then includes those bytes in the [`RawDocument`].
+//!
+//! ```rust
+//! use bson::{
+//!     raw::RawDocument,
+//!     doc,
+//! };
+//!
+//! let document = doc! {
```

Comment:
I don't remember if I got around to adding a `fn insert(&mut self, &str, Element)` method yet, but that would probably be a useful step for creating a rawdoc macro.  If it doesn't exist yet, a first version could just wrap doc!, though it would be less performant than direct insert.